### DISCUSSION TITLE
Constant Enums

### DISCUSSION THREAD

**Discussion by Panthr75** [dotnet/csharplang#6927] [Language Ideas] [answered]
After reading some discussion of #2849, I saw a lot of discussion regarding how enums aren't strict in terms of assigning arbitrary values to them.

Some enums do have fixed values. Take for example System.Threading.Tasks.TaskStatus

```cs
public const enum TaskStatus
{
  Created = 0,
  WaitingForActivation = 1,
  WaitingToRun = 2,
  Running = 3,
  WaitingForChildrenToComplete = 4,
  RanToCompletion = 5,
  Canceled = 6,
  Faulted = 7
}
```
by adding a const modifier, you let the compiler know that there can only be the numeric values 0, 1, 2, 3, 4, 5, 6, and 7 for Task status

Pros:
- Forces a strict structure

Cons:
- Fixed API. If I were to add a task status called `Interrupted`, now every consumer of this API could have their code break

Points of Interest:
- How would we handle adding a new `Interrupted` status if compiled code doesn't have a default check?
- Should we use `readonly` or `const` as the keyword?
- This could open up the door to Non ValueType members like string enums (#2849) and such by having them implicitly be constant by default.

**Comment by HaloFour** [comment, marked as answer]
See: #3179

The problem isn't really the compiler, the runtime offers no mechanism of enforcement, so the compiler has to assume that the enum type could have any value of the underlying integral type.

**Reply by DirkBoer** [reply]
Sorry for being ignorant about the difference between compiler and runtime related features, but wouldn't it be possible to throw an exception while casting from a number to the const enum, like any other invalid cast? 

This would mainly be applicable for edge cases like deserialization, but at least the rest of your code base can enjoy static code analysis.

Adding a new value to an existing Enum right now is NOT static typed and you don't know what code paths will break in a large codebase.

If you assume C# is meant to be a language for the next decade, you would think somewhere there should be a path to enable static code analysis on enum values too.

**Reply by DirkBoer** [reply]
It would provide static code analysis plus it would remove a lot of NotImplemented bloat everywhere:

```
if ( value == SomeEnum.ValueA )
	return XXX;
else if ( value == SomeEnum.ValueB )
	return YYY;
else
	throw new NotImplementedException($"{value} not implemented");
```
